### ISSUE TITLE
[Future feature] Allow generating JS API based migrations

### ISSUE THREAD

**Issue by dantman** [typeorm/typeorm#3607]
**Issue type:** [x] feature request
**Database system/driver:** n/a
**TypeORM version:** [x] `latest`

Right now using migrations:generate appears to generate SQL query migrations like:
```ts
await queryRunner.query("CREATE TABLE `tag` (`id` int NOT NULL AUTO_INCREMENT, `name` varchar(255) NOT NULL, PRIMARY KEY (`id`)) ENGINE=InnoDB");
```

It would be preferable if it were possible to automatically generate JS API based migrations. Using raw SQL queries for migrations has several disadvantages:

Create table queries for large new models can be long, I ended up with one line of a migration being 1KB long.

Raw SQL is not portable. TypeORM is relatively portable Entities are generic, queries are done in a database independent way, most types already have good enough default JS type -> db type mappings that you don't need to explicitly specify types, and when you do use explicitly typed columns they are defined in JS so it's possible that with some work you could define different types for different database engines.

But raw SQL migrations end up hardcoding database specific syntax and database specific types. So even if your models are database portable, your migrations are not. But JS API based migrations are about as portable as the models.

**Comment by nolazybits** [comment]
isn't it this?
https://github.com/typeorm/typeorm/blob/master/docs/migrations.md#using-migration-api-to-write-migrations

**Comment by dantman** [comment]
@nolazybits Yes, the API to write these manually does exist. The problem is that they are not generated automatically. TypeORM has a recommended `migrations:generate` command that generates code for a migration. But that generated code uses raw queries, it does not generate migrations that use the official migration API.

This request is basically one for `migrations:generate` to generate migrations that use that API, instead of generating raw SQL queries.

**Comment by nolazybits** [comment]
Yup that would be nice :)
My team just had to rewrite migration using the API for this exact reason. Generated SQL was specific to the targeted engine.

**Comment by Kononnable** [comment]
I see two more things to be considered.
 In pure sql we know exactly what will be done, in JS API it will be hidden under abstraction. So it might be harder to know if e.g. there is alter table statement underneath or drop and create statements. Also migrations by JS API can produce different schema when run on different typeorm versions. It's rarely but still possible(e.g. bug).

Maybe this should be implemented as an optional form of migration generation.

**Comment by PaulMest** [comment]
I see the concern with having SQL syntax obfuscated if this were changed to be more inline with other ORMs and migration generators. But I think the following will mitigate it:

* There should be logging for the migrations as they execute the real SQL that is generated.
* You could introduce a new CLI option to generate the SQL syntax.

For example, in Django, you can generate the SQL for a migrations by running:
`$ django-admin sqlmigrate`

This allows you to audit what is going to be run before it is run. And, if you wanted to, you could check in the SQL file next to the migration file.

So in typeorm, similar to how there is a `$ typeorm schema:log` there could be something like:

1. `$ typeorm migration:log`
2. `$ typeorm migration:run:log`
3. `$ typeorm migration:sql`

**Comment by Kleywalker** [comment]
Are there any updates yet? I really would like to generate TypeScript API Code.

**Comment by Kleywalker** [comment]
Is this being considered as a possible feature, and if so, will it be implemented in the near future?

**Comment by spotykatch** [comment]
> Is this being considered as a possible feature, and if so, will it be implemented in the near future?

Seems like, none of free js/ts orms are even thinking of this. You'll need a big team to support only this feature from version to version.

**Comment by nelson6e65** [comment]
Is there any good alternative to generate migrations in "API" syntax?
